docs/tutorials/Containers.md

```diff
@@ -0,0 +1,475 @@
+
+On local backends, you have to configure Cromwell to use a different `submit-docker` script that would start Singularity instead of docker.
+Singularity requires docker images to be prefixed with the prefix `docker://`.
+An example submit script for Singularity is:
+```bash
+singularity exec --bind ${cwd}:${docker_cwd} docker://${docker} ${job_shell} ${script}
+```
+
+As the Singularity container does not emit a job-id, we must include the `run-in-background` tag within the the provider section in addition to the docker-submit script. As Cromwell watches for the existence of the `rc` file, the `run-in-background` option has the caveat that we require the Singularity container to successfully complete, otherwise the workflow might hang indefinitely.
+
+Putting this together, we have an example base configuration for a local environment:
+```hocon
+include required(classpath("application"))
+
+backend {
+    default: singularity
+    providers: {
+        singularity {
+            # The backend custom configuration.
+            actor-factory = "cromwell.backend.impl.sfs.config.ConfigBackendLifecycleActorFactory"
+
+            config {
+                run-in-background = true
+                runtime-attributes = """
+                  String? docker
+                """
+                submit-docker = """
+                  singularity exec --bind ${cwd}:${docker_cwd} docker://${docker} ${job_shell} ${script}
+                """
+            }
+        }
+    }
+}
+```
+
+##### Job schedulers
+
+At its most basic level, if we want to run Singularity on a job scheduler, we need to pass the singularity command to the scheduler as a wrapped command.
+If we were using SLURM, this means we can use all the normal SLURM configuration as explained in the [SLURM documentation](../backends/SLURM). 
+However, our submit script will be a wrapped version of the script above:
+```
+submit-docker = """
+  # Ensure singularity is loaded if it's installed as a module
+  module load Singularity/3.0.1
+  
+  # Submit the script to SLURM
+  sbatch \
+    --wait \
+    -J ${job_name} \
+    -D ${cwd} \
+    -o ${cwd}/execution/stdout \
+    -e ${cwd}/execution/stderr \
+    -t ${runtime_minutes} \
+    ${"-c " + cpus} \
+    --mem-per-cpu=${requested_memory_mb_per_core} \
+    --wrap "singularity exec --bind ${cwd}:${docker_cwd} docker://${docker} ${job_shell} ${script}"
+  """
+```
+
+On some HPC systems, worker nodes do not have stable access to the internet or build access.
+If this is the case, you'll need to pull the Docker image before you actually submit the SLURM job.
+
+```
+submit-docker = """
+    [...]
+  
+    # Build the Docker image into a singularity image, using the head node
+    IMAGE=${cwd}/${docker}.sif
+    singularity build $IMAGE docker://${docker}
+  
+    # Submit the script to SLURM
+    sbatch \
+      [...]
+      --wrap "singularity exec --bind ${cwd}:${docker_cwd} $IMAGE ${job_shell} ${script}"
+  """
+```
```

Comment:
I would follow this up by saying "if worker nodes have access to a shared filesystem, it's best to build the container beforehand, and then provide a direct path to it"

(then put example of build here, followed by config with path). 

If the above is true, we are hugely helpful to advise the user to not run a thing that is going to try downloading the same layers to 1000 different worker nodes when it doesn't have to.

Reply:
I've added these three dot-points to give motivation (and reasoning) behind the build and execs, and removed one of the simpler configs as it's kind of not great practi_s_e given you probably don't want people to accidentally download and build the layers, and then consider the file system on top of that.

> - Make sure Singularity is loaded (and in our path), for example you can call module load Singularity/3.0.1.
> - We should treat worker nodes as if they do not have stable access to the internet or build access, so we will pull the container before execution of the task.
> - It's a good idea to ask Singularity to build the image into the execution directory of the task as an artifact, and to save rendering time on the worker node.

IMO, I think giving them the full context and config, and they want to strip it back to be _simpler_, then they can do that.

Reply:
This topic is a tricky one. Building the images once beforehand is obviously a better solution because it saves having to rebuild the image. But the point of Cromwell is that we want to run the exact same workflow on HPC as on Cloud. And if you have to manually build images and then edit the workflow in order to find that image file, then we're not facilitating that. At very least, Singularity will cache the Docker layers. As I've said before, if Cromwell had a `build-docker` or `pull-docker` hook that let us build once before running, that would solve the problem. But in the meantime I actually don't want to add the recommendation to build the container once beforehand, because it harms portability, increases the amount of work to run a portable workflow, and confuses this already confusing topic for readers.

Reply:
It's more than just saving time to build the images, for some clusters the concurrent downloads could hurt the performance of the entire system, and not just for one user. Hiding this information because Cromwell doesn't have a command line flag doesn't help anyone. The users want to run the job right, and it's our job to provide complete information. It's wrong for us to hide information from them for any reason. It's wrong for us to make decisions on their behalf that hide important details. It's reasonable to say "if you are able to do A, then do A" and "if you are not able to do A, then do B."

Reply:
I'll second this for udocker as well, udocker will (re)build the image on each run, there's a way to pull and build (_create_), but there doesn't seem to be an intuitive way to get the built image, or build otherwise. Every time you `udocker create` it will rebuild the image, and there's no easy way to look up by the floating tag.

Reply:
Okay, well then I'll write a section explaining the downside to this approach, and explain the alternative of pre-building images. But how would we actually implement this with Cromwell config? If they go through the whole workflow and build a bunch of `.sif` files into a directory, how do we now tell Cromwell to use them? Do we then have to also edit every task command line to point to these image files?

Reply:
@geoffjentry is this possible?

Reply:
I want to get this documentation in ASAP, so I hope we can find a workaround. However at the same time I wonder what long term changes need to be made to fix this problem properly. Ideally we want Singularity to work like Docker, which I think can handle multiple simultaneous pull commands. I've opened a Cromwell issue [here](https://github.com/broadinstitute/cromwell/issues/4673) that would solve it from one angle, but maybe this problem is better solved by better image caching in Singularity? What do you all think?

Reply:
From the [release notes](https://github.com/sylabs/singularity/releases) it looks like more work is being done for caching,  but we have to outline your use case. What Singularity already does is cache docker layers, and then assemble them at runtime. This applies to docker pulls. For shub (or image binary address) pulls, that's when you get an image file. So the strategy to get singularity support for caching isn't to use docker, but actual (entire) image binaries, either via some https address or their library.

Reply:
We still need the caching to work given a docker URL, because that's how all WDL/CWL workflows will be written. But we want Singularity to cache the whole image that it builds from that, not just the layers.

Reply:
I don't think that need is served, so you should open an issue and request it.